### ISSUE TITLE
Support env preparation&actual testing before and after the reboot

### ISSUE THREAD

**Issue by fernflower** [teemtee/tmt#694]
[Discussed in the tmt hacking session 30/03/21]

For testing the upgrade we need to run the actual test twice - first on the pristine not-yet-upgraded machine with rhel7, then on the very same machine after it's been upgraded to rhel8. In between the 2 test runs different playbooks are executed with reboot step.
A crude scenario:
Setup old distro (ansible roles/playbooks, reboot)
Test the application (check results)
Upgrade the distro (ansible roles/playbooks, reboot)
Test the application (check results)

The offered in the hacking session solution would work for us.
```
/upgrade:
  provision:
    how: virtual
    shared: true

  /old:
    prepare:
      how: ansible
      playbook: old.yaml
    execute:
      how: tmt
  /new:
    prepare:
      how: ansible
      playbook: new.yaml
    execute:
      how: tmt
```

**Comment by psss** [comment]
The original proposal above does not seem to be flexible enough. It would mean having two separate plans and thus reporting two results. Plus some other complications. Here's some further brainstorming of possible approaches:

Introduce a new `upgrade` step config which would be very similar to the `prepare` step but would be run between the old and new test execution:

```yaml
discover:
    how: fmf
prepare:
    how: ansible
    playbook: old.yaml
upgrade:
    how: ansible
    playbook: new.yaml
execute:
    how: tmt
```

Introduce a new `upgrade` method for test execution which would take care of all the special stuff for the upgrade testing:

```yaml
discover:
    how: fmf
prepare:
    how: ansible
    playbook: old.yaml
execute:
    how: upgrade
    playbook: new.yaml
```

A bit more general approach could be to allow multiple steps of the same name to be configured in the plan:

```yaml
discover:
    how: fmf
prepare-1: 
    how: ansible
    playbook: old.yaml
execute-1: 
    how: tmt
prepare-2:
    how: ansible
    playbook: old.yaml
execute-2:
    how: tmt
```

To properly handle the step order it would be probably better to use the `order` key:

```yaml
discover:
    how: fmf
    
prepare-old:
    how: ansible
    playbook: old.yaml
execute-old:
    how: tmt
    
prepare-new:
    how: ansible
    playbook: new.yaml
    order: 60
execute-new:
    how: tmt
    order: 70 
```

Each step would have some reasonable default, here's the full list to make the idea more tangible:

```yaml
discover:
    how: fmf
    order: 10
provision:
    how: virtual
    order: 20 
prepare-old:
    how: ansible
    playbook: old.yaml
    order: 40
execute-old:
    how: tmt
    order: 50
prepare-new:
    how: ansible
    playbook: new.yaml
    order: 60
execute-new:
    how: tmt
    order: 70
report:
    how: html
    order: 80
finish:
    how: shell
    order: 90
```

@fernflower, @ZhukovGreen, @lukaszachy, @sopos, @jscotka, @thrix, please have a look and add your brainstorming thoughts.

**Comment by ZhukovGreen** [comment]
@fernflower @psss 
can we just have two separate plans?
1. Just run the test suit on the initial OS (i.e. RHEL7)
2. Plan upgrades and reboots the machine during the preparation step and then run the same (or not the same) test suit on the upgraded machine

I believe prepare step is important to keep as exactly prepare (for the test). In this case, it makes the understanding of the 

Considering that the test should not modify the machine state, then why would you need to run those tests (before and after upgrade) in one plan?

**Comment by pvalena** [comment]
Posted my Idea here: https://github.com/psss/tmt/pull/701#issuecomment-817864920

**Comment by psss** [comment]
The main motivation for the single plan examples above is that there's just a single `discover` and `provision` step and also I understood this as a single scenario which should result into a single test result. Two separate plans sounds more like two different scenarios.

The upgrade testing use case has just two extra actions in addition to the standard six steps: Second `prepare` and second `execute`, the rest is the same:

* discover
* provision
* prepare
* execute
    * upgrade
    * execute
* report
* finish

As for now, injecting those two extra steps seems to me as the most straightforward way how to cover this scenario. But I see this is tailored very much to this particular use case. Not sure whether we really need the more generic approach which would cover various other combinations (also this would be much more complex to implement).

Let's continue with the brainstorm...

**Comment by fernflower** [comment]
> @fernflower @psss
> can we just have two separate plans?
I am afraid that is not an option - the result of the upgrade test run is the cumulative result of "run test before the upgrade, upgrade, run test after the upgrade". Splitting the run in 2 plans will only complicate things imho.

> I believe prepare step is important to keep as exactly prepare (for the test). In this case, it makes the understanding of the
Prepare steps for the same vm before and after the upgrade will be different.

> Considering that the test should not modify the machine state, then why would you need to run those tests (before and after upgrade) in one plan?
The test **will** modify the machine state, and running the same test before and after the upgrade on the very same vm is the definition of the upgrade tests we have for oamg.

> If the before upgrade tests modifies the system, then wouldn't it better to run as a prepare step?
This will mean that we will have to incorporate test on non-upgraded system&upgrade in the prepare step. That was the way I planned to take atm (while the discussed pre-\post- upgrade testing support is not in tmt yet) but in the long-term that is not the ideal solution imho.

**Comment by sopos** [comment]
another possibility with explicit ordering:
```
plan:
- discover:
- provision:
- prepare:
- execute:
- prepare:
  # do upgrade
- execute:
- report:
- finish:

```

**Comment by thrix** [comment]
And to make the feedback long, I have also an idea: I would like to introduce stages into execute. I believe for leapp, even the upgrade "stage" is a part of the test. Those stages should have generic names and different implementation (shell, ansible, etc.):

```
discover:
    how: fmf
provision:
    how: virtual
prepare:
    - name: preparation steps     # not part of the test execution, done only once
      how: ansible
      playbook: prepare.yml
execute:
    - name: reboot-for-upgrade   # generates its own test result (name needs to be unique)
      how: ansible
      playbook: reboot.yml
    - name: test via beakerlib      # list of tests comes from discover step
      how: tmt      
    - name: reboot-again             # generates its own test result  (name needs to be unique)
      how: ansible
      playbook: reboot.yml
    - name: test via shell             # generates its own test result  (name needs to be unique)
      how: shell
      script: echo hello
    - name: test via ansible         # generates its own test result  (name needs to be unique)
      how: ansible
      playbook: some-ansible-test.yml
report:
    how: html
finish:
    how: shell
    script: 
```

**Comment by fernflower** [comment]
> And to make the feedback long, I have also an idea: I would like to introduce stages into execute. I believe for leapp, even the upgrade "stage" is a part of the test. Those stages should have generic names and different implementation (shell, ansible, etc.):

FWIW I love the idea. This will need some careful way to report failures of the specific execute stage\non-trivial logs collection, but imho this solution is the closest to the oamg team's needs.

**Comment by psss** [comment]
I think @thrix has made an essential point above: that the `upgrade` stage is also part of the leapp test. I like this approach. There would be several execute configs (plugins), each with a unique name, the `report` step would gather results from all of them. This sounds like a way to go.

**Comment by ZhukovGreen** [comment]
@fernflower thanks for explaining this, I was not aware of such test scenarios. I like the proposal from @thrix . Just probably need to explain the usecase of execute:how clearly,

**Comment by psss** [comment]
Nice, so it seems we have an agreement on the general approach to this. @fernflower, @ZhukovGreen, are you interested to dive a little bit more deeper and work on the implementation of this new feature?

**Comment by fernflower** [comment]
> Nice, so it seems we have an agreement on the general approach to this. @fernflower, @ZhukovGreen, are you interested to dive a little bit more deeper and work on the implementation of this new feature?

@psss Sure thing, I'd love to work on the implementation.
There is also another feature from tmt that our upgrade testing verification will need, the "plan of plans" (To verify the build we actually run series of upgrade tests, say, a set of specific 21 tests with every test being run on a separate guest). I would also like to discuss and help with the implementation of that feature.

**Comment by ZhukovGreen** [comment]
I am sorry guys, but I need to finish first the integration of the libvirt and docker plugins and ensure our convert2rhel tests will be compatible with TFT.

**Comment by psss** [comment]
The upgrade testing scenario from the issue description should be covered by the new [upgrade](https://tmt.readthedocs.io/en/latest/spec/plans.html#upgrade) plugin for the `execute` step. Closing, feel free to reopen if anything essential is still missing.
